File: 백준/Silver/11060. 점프 점프/점프 점프.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;

		int N = Integer.parseInt(br.readLine());
		st = new StringTokenizer(br.readLine());
		int[] arr = new int[N];
		int[] dp = new int[1100];
		for (int i = 0; i < N; i++) {
			arr[i] = Integer.parseInt(st.nextToken());
			dp[i] = 1001;
		}

		dp[0] = 0;
		for (int i = 0; i < N; i++) {
			if (dp[i] == 1001)
				break;

			for (int j = 1; j <= arr[i]; j++) {
				dp[i + j] = Math.min(dp[i + j], dp[i] + 1);
			}
		}

		System.out.println(dp[N - 1] == 1001 ? -1 : dp[N - 1]);
	}
}

File: 백준/Silver/14248. 점프 점프/점프 점프.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.LinkedList
import java.util.StringTokenizer

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val n = readLine().toInt()
    val st = StringTokenizer(readLine())
    val arr = IntArray(n){st.nextToken().toInt()}
    val s = readLine().toInt()

    println(bfs(s, n, arr))
}

fun bfs(s: Int, max: Int, arr: IntArray): Int {
    var res = 1
    val visited = BooleanArray(max)
    val q = LinkedList<Int>()
    q.add(s - 1)
    visited[s - 1] = true

    while (!q.isEmpty()) {
        val now = q.poll()
        var next = now - arr[now]
        
        if (next >= 0 && !visited[next]) {
            res++
            visited[next] = true
            q.add(next)
        }

        next = now + arr[now]
        if (next < max && !visited[next]) {
            res++
            visited[next] = true
            q.add(next)
        }
    }

    return res
}